python/cuml/cuml/accel/profilers.py

```diff
@@ -0,0 +1,504 @@
+        _CALLBACKS.remove(self)
+        self.total_time += perf_counter() - self._start_time
+
+        if self.namespace is not None:
+            self.namespace.pop(self.FLAG, None)
+        sys.settrace(self._old_trace)
+
+        if not self.quiet:
+            self.print_report()
+
+    def _maybe_pop_timer(self):
+        if self._new_frame:
+            self._new_frame = False
+        else:
+            lineno, start = self._timers.pop()
+            duration = perf_counter() - start
+            stats = self.line_stats[lineno]
+            stats.count += 1
+            stats.total_time += duration
+
+    def _trace(self, frame, event, arg):
+        if not self._should_trace(frame):
+            # This frame is not directly in this script, no need to trace
+            return None
+
+        if event == "line":
+            # New line, maybe pop the last line's timer and push a new one.
+            self._maybe_pop_timer()
+            self._timers.append((frame.f_lineno, perf_counter()))
+        elif event == "return":
+            # Returning from a frame, maybe pop the last line's timer.
+            self._maybe_pop_timer()
+        elif event == "call":
+            # New function call, mark a new frame.
+            self._new_frame = True
+
+        return self._trace
+
+    def print_report(self) -> None:
+        """Print a report of the results."""
+        from rich.console import Console
+        from rich.style import Style
+        from rich.syntax import Syntax
+        from rich.table import Table
+
+        gpu_percent = 100 * self.total_gpu_time / self.total_time
+
+        table = Table(
+            title="cuml.accel line profile",
+            title_justify="left",
+            title_style=Style(bold=True),
+            caption=f"Ran in {format_duration(self.total_time)}, {gpu_percent:.1f}% on GPU",
+            caption_style=Style(),
+            caption_justify="left",
+        )
+        table.add_column("#", justify="right", no_wrap=True)
+        table.add_column("N", justify="right", no_wrap=True)
+        table.add_column("Time", justify="right", no_wrap=True)
+        table.add_column("GPU %", justify="right", no_wrap=True)
+        table.add_column("Source", justify="left")
+
+        for lineno, line in enumerate(self.source.splitlines(), 1):
+            stats = self.line_stats[lineno]
+
+            if stats.count == 0:
+                row = ("", "", "")
+            else:
+                if not stats.gpu_time and not stats.cpu_fallback:
+                    gpu_percent = "-"
+                elif stats.total_time > 0:
+                    gpu_percent = (
+                        f"{100 * stats.gpu_time // stats.total_time:.1f}"
+                    )
+                else:
+                    gpu_percent = "0.0"
+
+                if (
+                    not stats.gpu_time
+                    and not stats.cpu_fallback
+                    and stats.total_time < 0.001
```

Comment:
Rather than by the absolute total time, should we consider omitting results relative to the total runtime? Essentially drop all calls below the 0.1% percentile or so? Otherwise, we might have a scenario where everything is very fast and then everything is "-". This might not work as well in reality, so please feel free to reject this idea.

Reply:
I moved it to`min(point_one_percent_of_total_runtime, 0.001)`. I think this makes sense - python can do a lot in 1ms, and at that point the times might be of some relevant. But for very quick runs (<1s) we decrease the threshold so at least something still shows up. Here's an example on a very quick script:

```
cuml.accel line profile
┏━━━┳━━━┳━━━━━━━━┳━━━━━━━┳━━━━━━━━━━━━━━━━━━┓
┃ # ┃ N ┃   Time ┃ GPU % ┃ Source           ┃
┡━━━╇━━━╇━━━━━━━━╇━━━━━━━╇━━━━━━━━━━━━━━━━━━┩
│ 1 │ 1 │  3.3µs │     - │ a = 1            │
│ 2 │ 1 │  2.8µs │     - │ b = 2            │
│ 3 │ 1 │  2.4µs │     - │ c = 3            │
│ 4 │ 1 │ 25.6µs │     - │ print(a + b + c) │
└───┴───┴────────┴───────┴──────────────────┘
Ran in 439.9µs, 0.0% on GPU
```